htsinfer/get_read_layout.py

```diff
@@ -0,0 +1,254 @@
+        min_match_pct: float = 5,
+        min_freq_ratio: float = 2,
+    ):
+        """Class constructor."""
+        self.path: Path = path
+        self.adapter_file: Path = adapter_file
+        self.out_dir: Path = out_dir
+        self.min_match_pct: float = min_match_pct
+        self.min_freq_ratio: float = min_freq_ratio
+        self.adapters: List[Tuple[str, int]] = []
+        self.trie: ahc.Automaton = ahc.Automaton()
+        self.adapter_counts: DefaultDict[str, float] = defaultdict(lambda: 0)
+        self.result: Optional[str] = None
+
+    def evaluate(self) -> None:
+        """Search for adapter sequences and validate result confidence
+        constraints.
+        """
+        # load adapters
+        self._load_adapters()
+
+        # create trie of adapters
+        self._make_aho_auto()
+
+        records: int = 0
+        total_count: int = 0
+
+        try:
+            with open(self.path) as _f:  # type: ignore
+
+                LOGGER.debug("Procecssing Reads")
+                try:
+                    for record in FastqGeneralIterator(source=_f):
+                        read = record[1]
+
+                        # Searching for adapters in read
+                        for _, (_, key) in self.trie.iter(read):
+                            self.adapter_counts[key] += 1
+                            total_count += 1
+
+                        records += 1
+
+                except StopIteration as exc:
+                    self.result = None
+                    raise FileProblem(f"File is corrupt: {self.path}") from exc
+        except (OSError, ValueError) as exc:
+            self.result = None
+            raise FileProblem(f"{type(exc).__name__}: {str(exc)}") from exc
+
+        LOGGER.debug(f"Total records processed: {records}")
+
+        if total_count != 0:
+            # process and convert data to data frame
+            for i in self.adapter_counts:
+                self.adapter_counts[i] = round(
+                    (self.adapter_counts[i]/records)*100, 2
+                    )
+            adapters_df = self._convert_dict_to_df()
+
+            # write data frame (in JSON) to file
+            name = (
+                Path(self.out_dir) / f"read_layout_{Path(self.path).name}.json"
+            )
+            LOGGER.debug(f"Writing results to file: {name}")
+            adapters_df.to_json(
+                name,
+                orient='split',
+                index=False,
+                indent=True,
+            )
+
+            # validate results
+            if results_validator(
+                data=adapters_df,
+                column_index=1,
+                min_value=self.min_match_pct,
+                min_ratio=self.min_freq_ratio,
+            ):
+                self.result = adapters_df.iloc[0]['Adapter Sequence']
+            else:
+                self.result = None
+        else:
+            self.result = None
+
+    def _convert_dict_to_df(self) -> DataFrame:
```

Comment:
May actually be reused in other functionality so might be better placed as a static function in `utils.py`. I've moved the JSON file writing to the main method, but again, this might be reused, so consider moving to `utils.py`.